{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GSDC Raw Data Overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "glob.glob(\"./data/raw_data/*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_compounds_screened = pd.read_csv(\"data/raw_data/all_compounds_screened.csv\")\n",
    "all_cellines_screened = pd.read_excel(\"data/raw_data/all_cellines_screened.xlsx\", sheet_name=0)\n",
    "all_experiment = pd.read_excel(\"data/raw_data/GDSC2_drug_dose_cellines_IC50s.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 选取有测序结果和甲基化数据等数据的细胞系\n",
    "filtered_celline = all_cellines_screened.loc[\n",
    "    (all_cellines_screened['Whole Exome Sequencing (WES)'] == \"Y\") &\n",
    "    (all_cellines_screened['Methylation'] == \"Y\") &\n",
    "    (all_cellines_screened['Gene Expression'] == \"Y\") &\n",
    "    (all_cellines_screened['Copy Number Alterations (CNA)'] == \"Y\") &\n",
    "    (all_cellines_screened['Drug\\nResponse'] == \"Y\")\n",
    "]\n",
    "filtered_celline.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_celline.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = set(all_experiment['CELL_LINE_NAME'])\n",
    "b = set(filtered_celline['Sample Name'])\n",
    "print(f\"experiment sheet includes {a.__len__()} unqiue cellines\")\n",
    "print(f\"all_celline sheet includes {b.__len__()} unique cellines\")\n",
    "celline_barcode = list(a.intersection(b))\n",
    "print(f\"two sheets have {celline_barcode.__len__()} overlapping cellines\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GDSC Single Celline Multi-omics data Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNV data (Mutation Data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snv_df = pd.read_csv('data/raw_data/mutations_all_20230202.csv', low_memory=False)\n",
    "model_list = pd.read_csv('data/raw_data/model_list_20230306.csv', low_memory=False)\n",
    "snv_df.drop(columns='model_name', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snv_df = snv_df.join(model_list[['model_id', 'model_name']].set_index('model_id'), on=\"model_id\", how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snv_df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " important_only = ['cds_disrupted','complex_sub','downstream', 'ess_splice','frameshift','missense','nonsense','silent','splice_region','start_lost','stop_lost','upstream']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "'cds_disrupted', 终止密码子突变\n",
    " 'complex_sub', 复合物替换突变\n",
    " 'downstream', 下游\n",
    " 'ess_splice', 外显子剪辑沉默\n",
    " 'frameshift', 移码突变\n",
    " 'inframe', 开放框架内突变\n",
    " 'intronic', 内显子\n",
    " 'missense', 错义突变\n",
    " 'nc_ess_splice', \n",
    " 'nc_variant', \n",
    " 'nonsense', 无义突变\n",
    " 'silent', 沉默突变\n",
    " 'splice_region', 剪辑区域\n",
    " 'start_lost', 启动子丢失\n",
    " 'stop_lost', 终止密码子丢失\n",
    " 'upstream' 上游"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "important_only = ['cds_disrupted','complex_sub','downstream', 'ess_splice','frameshift','missense','nonsense','silent','splice_region','start_lost','stop_lost','upstream']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snv_df = snv_df[snv_df['effect'].isin(important_only)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_table = pd.pivot_table(data=snv_df, \n",
    "                          index='model_name', \n",
    "                          columns='gene_symbol', \n",
    "                          values='effect',\n",
    "                          aggfunc='count',\n",
    "                          fill_value=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methylation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "met_df = pd.read_csv('data/raw_data/F2_METH_CELL_DATA.txt', sep = '\\t', index_col=0)\n",
    "met_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "met_df = met_df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "met_lookup = pd.read_excel('data/raw_data/methSampleId_2_cosmicIds.xlsx', sheet_name=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "met_lookup['Sentrix_Barcode'] = list(\"_\".join([i,j]) for i, j in zip(met_lookup['Sentrix_ID'].astype(str), met_lookup['Sentrix_Position']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tb = dict(zip(met_lookup['Sentrix_Barcode'], met_lookup['Sample_Name']))\n",
    "met_df.rename(columns=tb, inplace=True)\n",
    "met_df = met_df.T"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy Number Variation and FPKM file readin "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cna = pd.read_csv(\"data/raw_data/celline_SNP6_cnv_gistics_20191101/cnv_abs_copy_number_picnic_20191101.csv\", low_memory=False,\n",
    "                      skiprows=lambda x: x in [0, 2])\n",
    "all_fpkm = pd.read_csv(\"./data/raw_data/cellines_rnaseq_all_20220624/rnaseq_fpkm_20220624.csv\", low_memory=False,\n",
    "                       skiprows=lambda x: x in [0, 2, 3, 4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(all_cna.drop(columns=['model_name', 'Unnamed: 1']).dropna().values)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Datasets Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_fpkm = all_fpkm.drop(columns=['model_name'])\n",
    "all_fpkm.rename(columns={\"Unnamed: 1\": \"celline_barcode\"}, inplace=True)\n",
    "all_fpkm.set_index('celline_barcode', inplace=True)\n",
    "all_fpkm = all_fpkm.T\n",
    "all_fpkm.fillna(0.0, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cna = all_cna.drop(columns=['model_name'])\n",
    "all_cna.rename(columns={\"Unnamed: 1\": \"celline_barcode\"}, inplace=True)\n",
    "all_cna.set_index('celline_barcode', inplace=True)\n",
    "all_cna = all_cna.T\n",
    "all_cna.fillna(0.0, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s1 = set(all_fpkm.index).intersection(celline_barcode)\n",
    "s2 = set(all_cna.index).intersection(celline_barcode)\n",
    "celline_barcode = list(s1.intersection(s2))\n",
    "print(f\"Two datasets have {len(celline_barcode)} common cellines\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_genes = list(set(cna_df.columns).intersection(fpkm_df.columns))\n",
    "print(f\"Two datasets have {len(common_genes)} genes(features)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_fpkm.loc[celline_barcode].to_csv(\"./data/processed_data/fpkm.csv\", sep='\\t')\n",
    "all_cna.loc[celline_barcode].to_csv(\"./data/processed_data/cna.csv\", sep='\\t')\n",
    "all_fpkm = all_fpkm.loc[celline_barcode]\n",
    "all_cna = all_cna.loc[celline_barcode]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GDSC Dataset"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drug_info: https://www.cancerrxgene.org/downloads/drug_data\n",
    "drug_df = pd.read_csv('./data/raw_data/drug_info.csv', sep=',')\n",
    "drug_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pubchem_id exclude non-numeric rows\n",
    "import re\n",
    "nonnumber = re.compile(r'\\D+')\n",
    "pubchem_id = list(set(drug_df['pubchem']))\n",
    "pubchem_id = [i.split(',')[0] if \",\" in i else i for i in pubchem_id]\n",
    "pubchem_id = [i for i in pubchem_id if re.findall(pattern=nonnumber, string=i).__len__()==0]\n",
    "drug_df = drug_df[drug_df['pubchem'].isin(pubchem_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_experiment.join(drug_df.set_index('drug_name'), on='DRUG_NAME', how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_experiment = all_experiment[all_experiment['CELL_LINE_NAME'].isin(celline_barcode)]\n",
    "all_experiment.reset_index(inplace = True)\n",
    "all_experiment.drop(columns = \"index\", inplace = True)\n",
    "all_experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_experiment.to_csv(\"./data/processed_data/expriment.csv\", sep=\"\\t\", index = None)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Response and Pubchem ID troubleshooting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pubchempy as pcp\n",
    "df = pcp.get_properties(properties=['canonical_smiles'], identifier=list(all_experiment['pubchem']),\n",
    "                        namespace='cid', )\n",
    "df = pd.DataFrame(df)\n",
    "df[['CID']]=df[['CID']].astype(str)\n",
    "df.to_csv(\"./data/processed_data/pubchem_id-SMILES.csv\", sep='\\t')\n",
    "lookup_table_cid_smiles = dict(zip(df['CID'], df['CanonicalSMILES']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_experiment = all_experiment[all_experiment['PUBCHEM_ID'].isin(pubchem_id)]\n",
    "all_experiment['SMILES']=[lookup_table_cid_smiles[i] for i in all_experiment['PUBCHEM_ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_experiment.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_barcode = [f\"{i[0]}_{i[1]}\" for i in zip(all_experiment['CELL_LINE_NAME'], all_experiment['PUBCHEM_ID'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "response = pd.DataFrame()\n",
    "response['sample_barcode'] = sample_barcode\n",
    "response['LN_IC50'] = all_experiment['LN_IC50']\n",
    "response.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**exclude outliers**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot\n",
    "ln_ic50 = all_experiment['LN_IC50'].values\n",
    "df = pd.DataFrame(ln_ic50)\n",
    "\n",
    "lower, median, upper = df.quantile([0.15,0.5,0.85]).values\n",
    "IQR = upper - lower\n",
    "lower_limit = lower - 1.5*IQR\n",
    "upper_limit = upper + 1.5*IQR\n",
    "\n",
    "all_experiment.loc[(all_experiment['LN_IC50'] < upper_limit.data) &\n",
    "                   (all_experiment['LN_IC50'] > lower_limit.data)]\n",
    "\n",
    "all_experiment.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "response.to_csv('./data/processed_data/response.csv', sep='\\t', index=None)\n",
    "all_experiment.to_csv('./data/processed_data/expriment.csv', sep='\\t', index=None)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drug Feature Extraction using beta-VAE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from os.path import realpath\n",
    "import pandas as pd\n",
    "from keras.models import load_model\n",
    "import ast\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from rdkit import Chem, RDLogger\n",
    "from rdkit.Chem import BondType\n",
    "\n",
    "RDLogger.DisableLog(\"rdApp.*\")\n",
    "\n",
    "SMILE_CHARSET = '[\"C\", \"B\", \"F\", \"I\", \"H\", \"O\", \"N\", \"S\", \"P\", \"Cl\", \"Br\"]'\n",
    "\n",
    "bond_mapping = {\"SINGLE\": 0, \"DOUBLE\": 1, \"TRIPLE\": 2, \"AROMATIC\": 3}\n",
    "bond_mapping.update(\n",
    "    {0: BondType.SINGLE, 1: BondType.DOUBLE, 2: BondType.TRIPLE, 3: BondType.AROMATIC}\n",
    ")\n",
    "SMILE_CHARSET = ast.literal_eval(SMILE_CHARSET)\n",
    "\n",
    "SMILE_to_index = dict((c, i) for i, c in enumerate(SMILE_CHARSET))\n",
    "index_to_SMILE = dict((i, c) for i, c in enumerate(SMILE_CHARSET))\n",
    "atom_mapping = dict(SMILE_to_index)\n",
    "atom_mapping.update(index_to_SMILE)\n",
    "\n",
    "BATCH_SIZE = 100\n",
    "EPOCHS = 10\n",
    "\n",
    "VAE_LR = 5e-4\n",
    "NUM_ATOMS = 120  # Maximum number of atoms\n",
    "\n",
    "ATOM_DIM = len(SMILE_CHARSET)  # Number of atom types\n",
    "BOND_DIM = 4 + 1  # Number of bond types\n",
    "LATENT_DIM = 435  # Size of the latent space\n",
    "\n",
    "\n",
    "def smiles_to_graph(smiles):\n",
    "    # Converts SMILES to molecule object\n",
    "    molecule = Chem.MolFromSmiles(smiles)\n",
    "\n",
    "    # Initialize adjacency and feature tensor\n",
    "    adjacency = np.zeros((BOND_DIM, NUM_ATOMS, NUM_ATOMS), \"float32\")\n",
    "    features = np.zeros((NUM_ATOMS, ATOM_DIM), \"float32\")\n",
    "\n",
    "    # loop over each atom in molecule\n",
    "    # Ignore Pt Atom\n",
    "    for atom in molecule.GetAtoms():\n",
    "        if(atom.GetSymbol() == \"Pt\"):\n",
    "            continue\n",
    "        i = atom.GetIdx()\n",
    "        atom_type = atom_mapping[atom.GetSymbol()]\n",
    "        features[i] = np.eye(ATOM_DIM)[atom_type]\n",
    "        # loop over one-hop neighbors\n",
    "        for neighbor in atom.GetNeighbors():\n",
    "            j = neighbor.GetIdx()\n",
    "            bond = molecule.GetBondBetweenAtoms(i, j)\n",
    "            bond_type_idx = bond_mapping[bond.GetBondType().name]\n",
    "            adjacency[bond_type_idx, [i, j], [j, i]] = 1\n",
    "\n",
    "    # Where no bond, add 1 to last channel (indicating \"non-bond\")\n",
    "    # Notice: channels-first\n",
    "    adjacency[-1, np.sum(adjacency, axis=0) == 0] = 1\n",
    "\n",
    "    # Where no atom, add 1 to last column (indicating \"non-atom\")\n",
    "    features[np.where(np.sum(features, axis=1) == 0)[0], -1] = 1\n",
    "\n",
    "    return adjacency, features\n",
    "\n",
    "\n",
    "def graph_to_molecule(graph):\n",
    "    # Unpack graph\n",
    "    adjacency, features = graph\n",
    "\n",
    "    # RWMol is a molecule object intended to be edited\n",
    "    molecule = Chem.RWMol()\n",
    "\n",
    "    # Remove \"no atoms\" & atoms with no bonds\n",
    "    keep_idx = np.where(\n",
    "        (np.argmax(features, axis=1) != ATOM_DIM - 1)\n",
    "        & (np.sum(adjacency[:-1], axis=(0, 1)) != 0)\n",
    "    )[0]\n",
    "    features = features[keep_idx]\n",
    "    adjacency = adjacency[:, keep_idx, :][:, :, keep_idx]\n",
    "\n",
    "    # Add atoms to molecule\n",
    "    for atom_type_idx in np.argmax(features, axis=1):\n",
    "        atom = Chem.Atom(atom_mapping[atom_type_idx])\n",
    "        _ = molecule.AddAtom(atom)\n",
    "\n",
    "    # Add bonds between atoms in molecule; based on the upper triangles\n",
    "    # of the [symmetric] adjacency tensor\n",
    "    (bonds_ij, atoms_i, atoms_j) = np.where(np.triu(adjacency) == 1)\n",
    "    for (bond_ij, atom_i, atom_j) in zip(bonds_ij, atoms_i, atoms_j):\n",
    "        if atom_i == atom_j or bond_ij == BOND_DIM - 1:\n",
    "            continue\n",
    "        bond_type = bond_mapping[bond_ij]\n",
    "        molecule.AddBond(int(atom_i), int(atom_j), bond_type)\n",
    "\n",
    "    # Sanitize the molecule; for more information on sanitization, see\n",
    "    # https://www.rdkit.org/docs/RDKit_Book.html#molecular-sanitization\n",
    "    flag = Chem.SanitizeMol(molecule, catchErrors=True)\n",
    "    # Let's be strict. If sanitization fails, return None\n",
    "    if flag != Chem.SanitizeFlags.SANITIZE_NONE:\n",
    "        return None\n",
    "\n",
    "    return molecule\n",
    "\n",
    "\n",
    "# 注：min-Max归一化需要在分割完训练集和测试集和Validation set之后再进行\n",
    "\n",
    "# Cached_data\n",
    "cached_data = {}\n",
    "\n",
    "# Path Define\n",
    "cna_path = realpath(\"data/processed_data/cna.csv\")\n",
    "experiment_path = realpath(\"data/processed_data/experiment.csv\")\n",
    "fpkm_path = realpath(\"data/processed_data/fpkm.csv\")\n",
    "SMILES_path = realpath(\"data/processed_data/pubchem_id-SMILES.csv\")\n",
    "\n",
    "drug_AdjacencyTensor = []\n",
    "drug_FeatureTensor = []\n",
    "\n",
    "df = pd.read_csv(SMILES_path, sep='\\t')\n",
    "for i in df[\"CanonicalSMILES\"]:\n",
    "    _ad, _fe = smiles_to_graph(i)\n",
    "    drug_AdjacencyTensor.append(_ad)\n",
    "    drug_FeatureTensor.append(_fe)\n",
    "\n",
    "drug_AdjacencyTensor = np.array(drug_AdjacencyTensor)\n",
    "drug_FeatureTensor = np.array(drug_FeatureTensor)\n",
    "\n",
    "vae = load_model(\"utils/drug-molecule-generation-with-VAE\",\n",
    "               compile=False)\n",
    "\n",
    "z_mean, _ = vae.encoder.predict([drug_AdjacencyTensor[:1000], drug_FeatureTensor[:1000]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['CID'].duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_feature_df = pd.DataFrame(data = z_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_feature_df.head(5) # 435-dim vector"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CNA 和 Gene-Expression dimension reduction"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CNA and FPKM data integration using SNF**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SNFtool: https://doi.org/10.1038/nmeth.2810\n",
    "\n",
    "R code is in current folder/SNL_integration.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!R CMD BATCH ./SNL_integration.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "similarity_df = pd.read_csv(\"data/processed_data/simlilarity_matrix.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "similarity_df.drop(columns=['Unnamed: 0'], inplace=True)\n",
    "similarity_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "celline_feature = {}\n",
    "for i, celline in enumerate(similarity_df.columns):\n",
    "    celline_feature[celline] = np.array(similarity_df.iloc[i].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = []\n",
    "for i in response['sample_barcode']:\n",
    "    celline_name, pubchem_id = i.split('_')\n",
    "    celline_feature_array = celline_feature[celline_name]\n",
    "    drug_feature_array = drug_feature_df.loc[int(pubchem_id)]\n",
    "    combined_feature = np.hstack([celline_feature_array, drug_feature_array])\n",
    "    s.append(combined_feature)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(s).shape # 866+435=1301-dim vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(\"./data/processed_data/all_feature\", np.array(s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = np.load('./data/processed_data/all_feature.npy')\n",
    "response = pd.read_csv('./data/processed_data/response.csv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "from tensorflow.data import Dataset\n",
    "from sklearn.preprocessing import MinMaxScaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = tf.constant(data)\n",
    "labels = response['LN_IC50'].values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scaler = MinMaxScaler()\n",
    "labels = labels.reshape(-1,1)\n",
    "scaler.fit(labels)\n",
    "labels = scaler.transform(labels)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Other Method\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "data = np.load('./data/processed_data/all_feature.npy')\n",
    "response = pd.read_csv('./data/processed_data/response.csv', sep='\\t')\n",
    "#features = tf.constant(data)\n",
    "labels = response['LN_IC50'].values\n",
    "scaler = MinMaxScaler()\n",
    "labels = labels.reshape(-1,1)\n",
    "scaler.fit(labels)\n",
    "labels = scaler.transform(labels)\n",
    "#labels = np.squeeze(1)\n",
    "#labels = tf.constant(labels)\n",
    "# dataset = Dataset.from_tensors((features, labels))\n",
    "\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.1, random_state=42)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "from tensorflow import keras\n",
    "from keras.losses import MeanSquaredError, BinaryCrossentropy\n",
    "from keras import Model\n",
    "from keras.metrics import Accuracy, AUC\n",
    "import pandas as pd\n",
    "from keras.models import Sequential\n",
    "from keras import layers\n",
    "import numpy as np\n",
    "from tensorflow.data import Dataset\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "data = np.load('./data/processed_data/all_feature.npy')\n",
    "response = pd.read_csv('./data/processed_data/response.csv', sep='\\t')\n",
    "#features = tf.constant(data)\n",
    "labels = response['LN_IC50'].values\n",
    "scaler = MinMaxScaler()\n",
    "labels = labels.reshape(-1,1)\n",
    "scaler.fit(labels)\n",
    "labels = scaler.transform(labels)\n",
    "# labels = np.squeeze(1)\n",
    "#labels = tf.constant(labels)\n",
    "# dataset = Dataset.from_tensors((features, labels))\n",
    "\n",
    "y = []\n",
    "for i in labels:\n",
    "    if (i[0]<=0.67):\n",
    "        y.append(1)\n",
    "    else:\n",
    "        y.append(0)\n",
    "labels = np.array(y)\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.33, \n",
    "                                                    random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "clf = RandomForestClassifier(max_depth=10, random_state=0)\n",
    "clf.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import f1_score, precision_score, recall_score\n",
    "y_pred = clf.predict(X_test)\n",
    "precision_score(y_test, y_pred)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "recall_score(y_test, y_pred)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MOFA analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from model.data import Dataset\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import muon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_data = Dataset()\n",
    "procssed_data = _data.return_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression = procssed_data['omics_data']['gene_expression']\n",
    "cnv = procssed_data['omics_data']['cnv']\n",
    "response = procssed_data['response']\n",
    "experiment = procssed_data['experiment']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_experiment = experiment[['SAMPLE_BARCODE', 'CELL_LINE_NAME', 'DRUG_NAME', 'pubchem', 'SMILES', 'AUC', 'Z_SCORE', 'LN_IC50']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_experiment"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare dataset with columns # The input needs to be a data.frame with columns [\"sample\",\"feature\",\"view\",\"group\",\"value\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_stakced = gene_expression.stack()\n",
    "cnv_stacked = cnv.stack()\n",
    "prepared_df = pd.DataFrame([[cll, symbol, 'gene_expression', '*', gene_expression_stakced[i]] for i, (cll, symbol) in enumerate(list(gene_expression_stakced.index))]+\n",
    "                           [[cll, symbol, 'cnv', '*', cnv_stacked[i]] for i, (cll, symbol) in enumerate(list(cnv_stacked.index))],\n",
    "                           columns=[\"sample\",\"feature\",\"view\",\"group\",\"value\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prepared_df.drop_duplicates([\"sample\",\"feature\",\"view\",\"group\"], keep = \"last\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mofapy2.run.entry_point import entry_point\n",
    "###########################\n",
    "## Initialise MOFA model ##\n",
    "###########################\n",
    "\n",
    "\n",
    "## (1) initialise the entry point\n",
    "ent = entry_point()\n",
    "## (2) Set data options\n",
    "# - scale_views: if views have very different ranges, one can to scale each view to unit variance\n",
    "ent.set_data_options(\n",
    "\tscale_views = False\n",
    ")\n",
    "\n",
    "# (3) Set data using the data frame format\n",
    "ent.set_data_df(prepared_df)\n",
    "\n",
    "# using default values\n",
    "ent.set_model_options()\n",
    "\n",
    "# using personalised values\n",
    "ent.set_model_options(\n",
    "\tfactors = 10, \n",
    "\tspikeslab_weights = True, \n",
    "\tard_weights = True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## (5) Set training options ##\n",
    "# - iter: number of iterations\n",
    "# - convergence_mode: \"fast\", \"medium\", \"slow\". Fast mode is usually good enough.\n",
    "# - dropR2: minimum variance explained criteria to drop factors while training. Default is None, inactive factors are not dropped during training\n",
    "# - gpu_mode: use GPU mode? this functionality needs cupy installed and a functional GPU, see https://biofam.github.io/MOFA2/gpu_training.html\n",
    "# - seed: random seed\n",
    "\n",
    "# using default values\n",
    "ent.set_train_options()\n",
    "\n",
    "# using personalised values\n",
    "ent.set_train_options(\n",
    "\titer = 100, \n",
    "\tconvergence_mode = \"fast\", \n",
    "\tdropR2 = None, \n",
    "\tgpu_mode = False, \n",
    "\tseed = 42\n",
    ")\n",
    "\n",
    "####################################\n",
    "## Build and train the MOFA model ##\n",
    "####################################\n",
    "\n",
    "# Build the model \n",
    "ent.build()\n",
    "\n",
    "# Run the model\n",
    "ent.run()\n",
    "\n",
    "####################\n",
    "## Save the model ##\n",
    "####################\n",
    "\n",
    "outfile = \"./test.hdf5\"\n",
    "\n",
    "# - save_data: logical indicating whether to save the training data in the hdf5 file.\n",
    "# this is useful for some downstream analysis in R, but it can take a lot of disk space.\n",
    "ent.save(outfile, save_data=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#########################\n",
    "## Downstream analysis ##\n",
    "#########################\n",
    "\n",
    "# Check the mofax package for the downstream analysis in Python: https://github.com/bioFAM/mofax\n",
    "# Check the MOFA2 R package for the downstream analysis in R: https://www.bioconductor.org/packages/release/bioc/html/MOFA2.html\n",
    "# All tutorials: https://biofam.github.io/MOFA2/tutorials.html\n",
    "\n",
    "# Extract factor values (a list with one matrix per sample group)\n",
    "factors = ent.model.nodes[\"Z\"].getExpectation()\n",
    "\n",
    "# Extract weights  (a list with one matrix per view)\n",
    "weights = ent.model.nodes[\"W\"].getExpectation()\n",
    "\n",
    "# Extract variance explained values\n",
    "r2 = ent.model.calculate_variance_explained()\n",
    "\n",
    "# Interact directly with the hdf5 file\n",
    "import h5py\n",
    "f = h5py.File(outfile, 'r')\n",
    "f.keys()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.heatmap(data=weights[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Extract factors\n",
    "# f[\"expectations\"][\"Z\"][\"*\"].value\n",
    "\n",
    "# Extract weights\n",
    "#f[\"expectations\"][\"W\"][\"gene_expression\"].value\n",
    "#f[\"expectations\"][\"W\"][\"cnv\"].value\n",
    "\n",
    "# Extract variance explained estimates\n",
    "print(f[\"variance_explained\"][\"r2_per_factor\"])\n",
    "print(f[\"variance_explained\"][\"r2_total\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv(\"data/processed_data/simlilarity_matrix.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(x = selected_experiment['LN_IC50'], bins=100)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CTRP Dataset Overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment = pd.read_csv(\"data/raw_data/CTRPv2.0/v20.data.curves_post_qc.txt\", sep='\\t')\n",
    "meta_celline = pd.read_csv(\"data/raw_data/CTRPv2.0/v20.meta.per_cell_line.txt\", sep=\"\\t\")\n",
    "meta_experiment = pd.read_csv(\"data/raw_data/CTRPv2.0/v20.meta.per_experiment.txt\", sep='\\t')\n",
    "meta_compound = pd.read_csv(\"data/raw_data/CTRPv2.0/v20.meta.per_compound.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_celline = meta_celline[['master_ccl_id', 'ccl_name']]\n",
    "meta_compound = meta_compound[['master_cpd_id', 'cpd_name', 'cpd_smiles']]\n",
    "meta_experiment = meta_experiment[['experiment_id','master_ccl_id']]\n",
    "experiment = experiment[['experiment_id', 'area_under_curve', 'master_cpd_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment = experiment.join(meta_experiment.set_index('experiment_id'), on='experiment_id',how='inner')\n",
    "experiment = experiment.join(meta_compound.set_index('master_cpd_id'), on='master_cpd_id',how='inner')\n",
    "experiment = experiment.join(meta_celline.set_index('master_ccl_id'), on='master_ccl_id', how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment = experiment[['experiment_id', 'master_cpd_id', 'cpd_name', 'cpd_smiles',\n",
    "                         'master_ccl_id', 'ccl_name', 'area_under_curve']] # reorder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from model.data import Dataset\n",
    "from config_path import *\n",
    "\n",
    "ds = Dataset(response='AUC', dataset='CTRP')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "查看AUC和LN_IC50之间的相关性"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat = ds.experiment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat['AUC'] = (dat['AUC'] - dat['AUC'].min()) / (dat['AUC'].max() - dat['AUC'].min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower, median, upper = dat[['AUC']].quantile([0.33,0.5,0.66]).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(lower)\n",
    "print(median)\n",
    "print(upper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.hist(ds.processed_experiment['AUC'], bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import keras\n",
    "import keras.backend as K\n",
    "import tensorflow as tf\n",
    "\n",
    "a = K._to_tensor(np.array([[1,2,3,4,5],[3,4,5,6,7]]), dtype=tf.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrp_drug_df = pd.read_csv('data/raw_data/CTRPv2.0/v20.meta.per_compound.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrp_drug_df['cpd_smiles']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd\n",
    "from config_path import *\n",
    "\n",
    "def smiles2canonical(smiles):\n",
    "    url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/{smiles}/property/CanonicalSMILES/JSON\"\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        d = json.loads(response.content)\n",
    "        print(d)\n",
    "        cid = d['PropertyTable']['Properties'][0].get('CID', None)\n",
    "        smiles = d['PropertyTable']['Properties'][0].get('CanonicalSMILES', None)\n",
    "        return (cid, smiles)\n",
    "    else:\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_smiles2canonical():\n",
    "    ctrp_drug_df = pd.read_csv(CTRP_DRUG_PATH, sep='\\t').astype(str)\n",
    "    smiles_list = []\n",
    "    cid_list = []\n",
    "    df = pd.DataFrame()\n",
    "    for idx, (name, smiles) in enumerate(list(zip(ctrp_drug_df['cpd_name'], ctrp_drug_df['cpd_smiles']))):\n",
    "        r = smiles2canonical(smiles)\n",
    "        if r is not None: \n",
    "            print(f\"Drug has corresponding canonical smiles and cid\")\n",
    "            smiles_list.append(r[1])\n",
    "            cid_list.append(r[0])\n",
    "        else:\n",
    "            print(f\"Drug has no corresponding Canonical SMILES.\")\n",
    "            smiles_list.append(None)\n",
    "            cid_list.append(None)\n",
    "    df['CanonicalSMILES'] = smiles_list\n",
    "    df.index = ctrp_drug_df['cpd_name']\n",
    "    df['CID'] = [str(i) for i in cid_list]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv('data/raw_data/screened_compounds_rel_8.4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from model.data import Dataset\n",
    "\n",
    "ctrp_ds = Dataset(dataset='CTRP')\n",
    "gdsc_ds = Dataset(dataset=\"GDSC\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test Hold-out dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Copy Number Abberation Data...\n",
      "Loading Copy Number Abberation Data Done\n",
      "Loading Gene Expression Data...\n",
      "Loading Gene Expression Data Done\n",
      "Loading Methylation Data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sunzehui/miniconda3/envs/tf/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Methylation Data Done\n",
      "Loading Mutations Data...\n",
      "Loading Mutations Data Done\n",
      "Begin loading drug data...\n",
      "After filtering, GDSC has tested 477 drugs; CTRP has tested 425 drugs\n",
      "After combining, unique cid number is 792\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sunzehui/miniconda3/envs/tf/lib/python3.10/site-packages/scipy/_lib/_util.py:69: RuntimeWarning: overflow encountered in _ncx2_cdf\n",
      "  np.place(out, cond, f(*temp))\n",
      "/Users/sunzehui/miniconda3/envs/tf/lib/python3.10/site-packages/scipy/stats/_continuous_distns.py:6630: RuntimeWarning: overflow encountered in _ncf_cdf\n",
      "  return _boost._ncf_cdf(x, dfn, dfd, nc)\n",
      "/Users/sunzehui/miniconda3/envs/tf/lib/python3.10/site-packages/scipy/stats/_continuous_distns.py:6826: RuntimeWarning: overflow encountered in _nct_cdf\n",
      "  return np.clip(_boost._nct_cdf(x, df, nc), 0, 1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drug data loaded\n",
      "Loading CTRP Experiment Data...\n",
      "Loading Experiment Data Done\n",
      "Begin Preprocessing Experiment!\n",
      "Select Overlapping Cellines...\n",
      "Index(['DRUG_NAME', 'SMILES', 'CELL_LINE_NAME', 'AUC', 'DATASET'], dtype='object')\n",
      "Select Overlapping Cellines with available PubChem CIDs...\n",
      "Create Unique Sample Barcode...\n",
      "Exclude response value...\n",
      "Experiment Done!\n",
      "Preparing Omics data...\n",
      "Omics data Done!\n",
      "Save the dataset into hdf5 data format...\n",
      "Done!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-04-16 21:21:36.444965: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/3\n",
      "1161/1161 [==============================] - 237s 202ms/step - loss: 0.5183 - precision: 0.5674 - recall: 0.7621 - auc: 0.8279 - auc_1: 0.6835 - val_loss: 0.4643 - val_precision: 0.6094 - val_recall: 0.8883 - val_auc: 0.9040 - val_auc_1: 0.8276 - lr: 0.0010\n",
      "Epoch 2/3\n",
      "1161/1161 [==============================] - 239s 206ms/step - loss: 0.4176 - precision: 0.6514 - recall: 0.8217 - auc: 0.8907 - auc_1: 0.7930 - val_loss: 0.4022 - val_precision: 0.6596 - val_recall: 0.8856 - val_auc: 0.9156 - val_auc_1: 0.8469 - lr: 0.0010\n",
      "Epoch 3/3\n",
      "1161/1161 [==============================] - 243s 209ms/step - loss: 0.4018 - precision: 0.6652 - recall: 0.8259 - auc: 0.8993 - auc_1: 0.8106 - val_loss: 0.3775 - val_precision: 0.6858 - val_recall: 0.8594 - val_auc: 0.9198 - val_auc_1: 0.8519 - lr: 0.0010\n",
      "145/145 [==============================] - 27s 184ms/step - loss: 0.3675 - precision: 0.7014 - recall: 0.8642 - auc: 0.9213 - auc_1: 0.8598\n",
      "[0.36745336651802063, 0.7013773918151855, 0.8642226457595825, 0.9212973713874817, 0.8598474860191345]\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "from keras.metrics import AUC\n",
    "from keras.callbacks import LearningRateScheduler, ReduceLROnPlateau, EarlyStopping\n",
    "from keras.metrics import Precision, Recall\n",
    "from keras.losses import BinaryCrossentropy\n",
    "import datetime\n",
    "from model.nn import multichannel_network\n",
    "from model.data import Dataset, DataGenerator\n",
    "from sklearn.utils import class_weight\n",
    "import numpy as np\n",
    "\n",
    "# Dataset Setting: \n",
    "## choose from ['methylation', 'gene_expression', 'cnv', 'mutation']\n",
    "FEATURE = ['gene_expression', 'cnv', 'methylation', 'mutation']\n",
    "ds = Dataset(\n",
    "    feature_contained=FEATURE, \n",
    "    dataset='CTRP', \n",
    "    set_label=True, \n",
    "    response='AUC', \n",
    "    threshold=.58)\n",
    "# CTRP, \"AUC\", 0.58, 0.001\n",
    "# GDSC, \"AUC\", .88, 0.001\n",
    "# model parameters settings\n",
    "lr_rate = 0.001\n",
    "dropout_rate = .5\n",
    "batch_size = 32 \n",
    "epochs = 3\n",
    "\n",
    "# Split train, test and validation set for training and testing, build generators\n",
    "partition = ds.split(validation=True)\n",
    "train = partition['train']\n",
    "test = partition['test']\n",
    "validation = partition['validation']\n",
    "train_generator = DataGenerator(sample_barcode=train, **ds.get_config(), batch_size=batch_size)\n",
    "validation_generator = DataGenerator(sample_barcode=validation, **ds.get_config(), batch_size=batch_size)\n",
    "test_generator = DataGenerator(sample_barcode=test, **ds.get_config(), batch_size=batch_size)\n",
    "\n",
    "# Training parameters\n",
    "class_weights = class_weight.compute_class_weight(class_weight='balanced',\n",
    "                                                 classes=np.unique([ds.labels[x] for x in train]),\n",
    "                                                 y=[ds.labels[x] for x in train])\n",
    "weights_dict = {i:w for i,w in enumerate(class_weights)}\n",
    "log_dir = \"logs/fit/\" + datetime.datetime.now().strftime(\"%Y%m%d-%H%M%S\")\n",
    "tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)\n",
    "def scheduler(epoch, lr):\n",
    "    if(epoch % 5 ==0 and epoch !=0):\n",
    "        return lr*0.1\n",
    "    else:\n",
    "        return lr\n",
    "reduce_lr = LearningRateScheduler(scheduler)\n",
    "# reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2,\n",
    "#                               patience=5, min_lr=0.001)\n",
    "early_stop = EarlyStopping(monitor='val_loss', patience=10)\n",
    "\n",
    "\n",
    "# model building\n",
    "model = multichannel_network(\n",
    "    dataset=ds,\n",
    "    train_sample_barcode=train,\n",
    "    dropout=dropout_rate\n",
    "    )\n",
    "\n",
    "model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=lr_rate),\n",
    "              loss=BinaryCrossentropy(),\n",
    "              metrics=\n",
    "              [\n",
    "                Precision(name=\"precision\"),\n",
    "                Recall(name=\"recall\"),\n",
    "                AUC(curve='ROC'),\n",
    "                AUC(curve='PR')\n",
    "              ]\n",
    "            )\n",
    "\n",
    "history = model.fit(\n",
    "    x=train_generator, \n",
    "    epochs=epochs,\n",
    "    validation_data=validation_generator, \n",
    "    callbacks=[reduce_lr, early_stop],\n",
    "    class_weight=weights_dict\n",
    "                    )\n",
    "\n",
    "scores = model.evaluate(x=test_generator) \n",
    "print(list(scores))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_list = pd.read_csv('data/raw_data/model_list_20230307.csv')[['model_name', 'tissue', 'cancer_type']]\n",
    "colo_ccl = set(model_list[model_list['cancer_type'] == \"Colorectal Carcinoma\"]['model_name'])\n",
    "celline_barcode = set(ds.omics_data['cnv'].index)\n",
    "colo_ccl = colo_ccl.intersection(celline_barcode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HT55',\n",
       " 'KM12',\n",
       " 'MDST8',\n",
       " 'RKO',\n",
       " 'SW1116',\n",
       " 'SW1417',\n",
       " 'SW1463',\n",
       " 'SW48',\n",
       " 'SW948',\n",
       " 'T84'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colo_ccl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_candidate = list(ds.drug_info.all_drugs.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "experiment_candidate = itertools.product(colo_ccl, drug_candidate)\n",
    "experiment_candidate = [i for i in experiment_candidate]\n",
    "celline_candidate = [i[0] for i in experiment_candidate]\n",
    "drug_candidate = [i[1] for i in experiment_candidate]\n",
    "feature = {}\n",
    "for i in ds.feature_contained:\n",
    "    if i == \"cnv\":\n",
    "        feature['cnv'] = ds.omics_data['cnv'].loc[celline_candidate].values.astype(np.float32)\n",
    "    elif i == \"gene_expression\":\n",
    "        feature['gene_expression'] = ds.omics_data['gene_expression'].loc[celline_candidate].values.astype(np.float32)\n",
    "    elif i == \"mutation\":\n",
    "        feature['mutation'] = ds.omics_data['mutation'].loc[celline_candidate].values.astype(np.float32)\n",
    "    elif i == \"methylation\":\n",
    "        feature['methylation'] = ds.omics_data['methylation'].loc[celline_candidate].values.astype(np.float32)\n",
    "feature['fingerprint'] = ds.drug_info.drug_feature['fingerprint'].loc[drug_candidate].values.astype(np.float32)\n",
    "feature['rdkit2d'] = ds.drug_info.drug_feature['rdkit2d'].loc[drug_candidate].values.astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunks = []\n",
    "for i in range(0, 7920, 32):\n",
    "    x = i\n",
    "    chunks.append({\n",
    "        'cnv': feature['cnv'][x:x+32],\n",
    "        'gene_expression': feature['gene_expression'][x:x+32],\n",
    "        'methylation': feature['methylation'][x:x+32],\n",
    "        'mutation': feature['mutation'][x:x+32],\n",
    "        'fingerprint': feature['fingerprint'][x:x+32],\n",
    "        'rdkit2d': feature['rdkit2d'][x:x+32]\n",
    "    })\n",
    "# last chunk\n",
    "if len(experiment_candidate) % 32 != 0:\n",
    "    last_chunk = {}\n",
    "    leftover = len(experiment_candidate) % 32 \n",
    "    for i,j in feature.items():\n",
    "        last_chunk[i] = np.zeros(shape=(32, j.shape[1]))\n",
    "        last_chunk[i][0:leftover] = j[-leftover::]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n"
     ]
    }
   ],
   "source": [
    "result = []\n",
    "for idx, i in enumerate(chunks[:-1]):\n",
    "    result.append(model(i))\n",
    "result.append(model(last_chunk))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "AxisError",
     "evalue": "axis -2 is out of bounds for array of dimension 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAxisError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m result \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39;49mconcatenate(result, axis\u001b[39m=\u001b[39;49m\u001b[39m-\u001b[39;49m\u001b[39m2\u001b[39;49m)\n\u001b[1;32m      2\u001b[0m result \u001b[39m=\u001b[39m result[\u001b[39m0\u001b[39m:\u001b[39mlen\u001b[39m(experiment_candidate)]\n",
      "File \u001b[0;32m<__array_function__ internals>:180\u001b[0m, in \u001b[0;36mconcatenate\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;31mAxisError\u001b[0m: axis -2 is out of bounds for array of dimension 1"
     ]
    }
   ],
   "source": [
    "result = np.concatenate(result, axis=-2)\n",
    "result = result[0:len(experiment_candidate)]\n",
    "df = pd.DataFrame(data=result, columns=['AUC_predicted'])\n",
    "df['DRUG_NAME'] = drug_candidate\n",
    "df['CELL_LINE'] = celline_candidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exclued = df[~df['DRUG_NAME'].isin(ds.processed_experiment['DRUG_NAME'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_experiment = ds.experiment\n",
    "all_experiment['AUC'] = (all_experiment['AUC'] - all_experiment['AUC'].min())/(all_experiment['AUC'].max()-all_experiment['AUC'].min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'16-beta-bromoandrosterone',\n",
       " '1S,3R-RSL-3',\n",
       " '3-Cl-AHPC',\n",
       " '968',\n",
       " 'ABT-199',\n",
       " 'ABT-737',\n",
       " 'AC55649',\n",
       " 'AM-580',\n",
       " 'AT-406',\n",
       " 'AT13387',\n",
       " 'AT7867',\n",
       " 'AZ-3146',\n",
       " 'AZD1480',\n",
       " 'AZD4547',\n",
       " 'AZD6482',\n",
       " 'AZD7545',\n",
       " 'AZD7762',\n",
       " 'AZD8055',\n",
       " 'B02',\n",
       " 'BEC',\n",
       " 'BI-2536',\n",
       " 'BIBR-1532',\n",
       " 'BIRB-796',\n",
       " 'BIX-01294',\n",
       " 'BMS-195614',\n",
       " 'BMS-270394',\n",
       " 'BMS-345541',\n",
       " 'BMS-536924',\n",
       " 'BMS-754807',\n",
       " 'BRD-A02303741',\n",
       " 'BRD-A05715709',\n",
       " 'BRD-A71883111',\n",
       " 'BRD-A86708339',\n",
       " 'BRD-A94377914',\n",
       " 'BRD-K01737880',\n",
       " 'BRD-K02251932',\n",
       " 'BRD-K02492147',\n",
       " 'BRD-K03536150',\n",
       " 'BRD-K04800985',\n",
       " 'BRD-K07442505',\n",
       " 'BRD-K09344309',\n",
       " 'BRD-K09587429',\n",
       " 'BRD-K11533227',\n",
       " 'BRD-K16147474',\n",
       " 'BRD-K17060750',\n",
       " 'BRD-K19103580',\n",
       " 'BRD-K24690302',\n",
       " 'BRD-K26531177',\n",
       " 'BRD-K27986637',\n",
       " 'BRD-K28456706',\n",
       " 'BRD-K29086754',\n",
       " 'BRD-K29313308',\n",
       " 'BRD-K30019337',\n",
       " 'BRD-K30748066',\n",
       " 'BRD-K33199242',\n",
       " 'BRD-K34099515',\n",
       " 'BRD-K34485477',\n",
       " 'BRD-K37390332',\n",
       " 'BRD-K41334119',\n",
       " 'BRD-K42260513',\n",
       " 'BRD-K44224150',\n",
       " 'BRD-K45681478',\n",
       " 'BRD-K48334597',\n",
       " 'BRD-K48477130',\n",
       " 'BRD-K49290616',\n",
       " 'BRD-K50799972',\n",
       " 'BRD-K51490254',\n",
       " 'BRD-K51831558',\n",
       " 'BRD-K52037352',\n",
       " 'BRD-K55116708',\n",
       " 'BRD-K61166597',\n",
       " 'BRD-K64610608',\n",
       " 'BRD-K66453893',\n",
       " 'BRD-K66532283',\n",
       " 'BRD-K70511574',\n",
       " 'BRD-K71781559',\n",
       " 'BRD-K75293299',\n",
       " 'BRD-K78574327',\n",
       " 'BRD-K80183349',\n",
       " 'BRD-K84807411',\n",
       " 'BRD-K88742110',\n",
       " 'BRD-K90370028',\n",
       " 'BRD-K92856060',\n",
       " 'BRD-K96431673',\n",
       " 'BRD-K96970199',\n",
       " 'BRD-K99006945',\n",
       " 'BRD1812',\n",
       " 'BRD4132',\n",
       " 'BRD6340',\n",
       " 'BRD8899',\n",
       " 'BRD8958',\n",
       " 'BRD9647',\n",
       " 'BYL-719',\n",
       " 'Bax channel blocker',\n",
       " 'C6-ceramide',\n",
       " 'CAL-101',\n",
       " 'CAY10594',\n",
       " 'CAY10618',\n",
       " 'CBB-1007',\n",
       " 'CCT036477',\n",
       " 'CD-1530',\n",
       " 'CD-437',\n",
       " 'CHIR-99021',\n",
       " 'CHM-1',\n",
       " 'CI-976',\n",
       " 'CID-5951923',\n",
       " 'CIL41',\n",
       " 'CIL55',\n",
       " 'CIL56',\n",
       " 'CIL70',\n",
       " 'COL-3',\n",
       " 'CR-1-31B',\n",
       " 'Ch-55',\n",
       " 'Compound 1541A',\n",
       " 'Compound 23 citrate',\n",
       " 'DBeQ',\n",
       " 'ELCPK',\n",
       " 'EX-527',\n",
       " 'FGIN-1-27',\n",
       " 'FQI-1',\n",
       " 'FQI-2',\n",
       " 'GANT-61',\n",
       " 'GDC-0941',\n",
       " 'GSK-J4',\n",
       " 'GSK1059615',\n",
       " 'GSK2636771',\n",
       " 'GSK4112',\n",
       " 'GSK461364',\n",
       " 'GSK525762A',\n",
       " 'GW-405833',\n",
       " 'GW-843682X',\n",
       " 'HC-067047',\n",
       " 'HLI 373',\n",
       " 'I-BET151',\n",
       " 'IC-87114',\n",
       " 'IPR-456',\n",
       " 'ISOX',\n",
       " 'IU1',\n",
       " 'JQ-1',\n",
       " 'JW-480',\n",
       " 'JW-55',\n",
       " 'JW-74',\n",
       " 'KHS101',\n",
       " 'KU 0060648',\n",
       " 'KU-0063794',\n",
       " 'KU-55933',\n",
       " 'KU-60019',\n",
       " 'KW-2449',\n",
       " 'KX2-391',\n",
       " 'Ki8751',\n",
       " 'Ko-143',\n",
       " 'L-685458',\n",
       " 'LBH-589',\n",
       " 'LE-135',\n",
       " 'LRRK2-IN-1',\n",
       " 'LY-2157299',\n",
       " 'LY-2183240',\n",
       " 'MG-132',\n",
       " 'MGCD-265',\n",
       " 'MI-1',\n",
       " 'MI-2',\n",
       " 'MK-0752',\n",
       " 'MK-1775',\n",
       " 'MK-2206',\n",
       " 'ML006',\n",
       " 'ML029',\n",
       " 'ML031',\n",
       " 'ML050',\n",
       " 'ML083',\n",
       " 'ML162',\n",
       " 'ML203',\n",
       " 'ML210',\n",
       " 'ML239',\n",
       " 'ML311',\n",
       " 'ML312',\n",
       " 'ML320',\n",
       " 'ML334 diastereomer',\n",
       " 'MLN2238',\n",
       " 'MLN2480',\n",
       " 'MST-312',\n",
       " 'Mdivi-1',\n",
       " 'Merck60',\n",
       " 'N9-isopropylolomoucine',\n",
       " 'NPC-26',\n",
       " 'NSC 74859',\n",
       " 'NSC19630',\n",
       " 'NSC23766',\n",
       " 'NSC30930',\n",
       " 'NSC48300',\n",
       " 'NSC632839',\n",
       " 'NSC95397',\n",
       " 'NVP-231',\n",
       " 'NVP-ADW742',\n",
       " 'NVP-BSK805',\n",
       " 'NVP-TAE684',\n",
       " 'O-6-benzylguanine',\n",
       " 'OSI-027',\n",
       " 'OSI-930',\n",
       " 'PAC-1',\n",
       " 'PD 153035',\n",
       " 'PD318088',\n",
       " 'PDMP',\n",
       " 'PF-184',\n",
       " 'PF-3758309',\n",
       " 'PF-4800567 hydrochloride',\n",
       " 'PF-543',\n",
       " 'PF-573228',\n",
       " 'PF-750',\n",
       " 'PHA-793887',\n",
       " 'PI-103',\n",
       " 'PIK-93',\n",
       " 'PL-DI',\n",
       " 'PLX-4032',\n",
       " 'PLX-4720',\n",
       " 'PRIMA-1',\n",
       " 'PRIMA-1-Met',\n",
       " 'PRL-3 inhibitor I',\n",
       " 'PX-12',\n",
       " 'PYR-41',\n",
       " 'Platin',\n",
       " 'QS-11',\n",
       " 'QW-BI-011',\n",
       " 'R428',\n",
       " 'RAF265',\n",
       " 'RG-108',\n",
       " 'RITA',\n",
       " 'RO4929097',\n",
       " 'Repligen 136',\n",
       " 'SB-225002',\n",
       " 'SB-431542',\n",
       " 'SB-525334',\n",
       " 'SB-743921',\n",
       " 'SCH-529074',\n",
       " 'SCH-79797',\n",
       " 'SGX-523',\n",
       " 'SID 26681509',\n",
       " 'SKI-II',\n",
       " 'SMER-3',\n",
       " 'SN-38',\n",
       " 'SNS-032',\n",
       " 'SNX-2112',\n",
       " 'SR-II-138A',\n",
       " 'SR1001',\n",
       " 'SR8278',\n",
       " 'SRT-1720',\n",
       " 'STF-31',\n",
       " 'SU11274',\n",
       " 'SZ4TA2',\n",
       " 'StemRegenin 1',\n",
       " 'TG-100-115',\n",
       " 'TG-101348',\n",
       " 'TGX-221',\n",
       " 'TPCA-1',\n",
       " 'TW-37',\n",
       " 'UNC0321',\n",
       " 'UNC0638',\n",
       " 'VAF-347',\n",
       " 'VU0155056',\n",
       " 'VX-680',\n",
       " 'WAY-362450',\n",
       " 'WZ4002',\n",
       " 'WZ8040',\n",
       " 'XL765',\n",
       " 'YK 4-279',\n",
       " 'YM-155',\n",
       " 'ZSTK474',\n",
       " 'abiraterone',\n",
       " 'afatinib',\n",
       " 'alisertib',\n",
       " 'alvocidib',\n",
       " 'apicidin',\n",
       " 'austocystin D',\n",
       " 'avicin D',\n",
       " 'avrainvillamide',\n",
       " 'axitinib',\n",
       " 'azacitidine',\n",
       " 'bafilomycin A1',\n",
       " 'barasertib',\n",
       " 'belinostat',\n",
       " 'bendamustine',\n",
       " 'betulinic acid',\n",
       " 'bexarotene',\n",
       " 'blebbistatin',\n",
       " 'bleomycin A2',\n",
       " 'bortezomib',\n",
       " 'brivanib',\n",
       " 'cabozantinib',\n",
       " 'canertinib',\n",
       " 'carboplatin',\n",
       " 'cediranib',\n",
       " 'ceranib-2',\n",
       " 'cerulenin',\n",
       " 'chlorambucil',\n",
       " 'ciclopirox',\n",
       " 'ciclosporin',\n",
       " 'clofarabine',\n",
       " 'compound 1B',\n",
       " 'crizotinib',\n",
       " 'crizotinib:PLX-4032 (2:1 mol/mol)',\n",
       " 'cucurbitacin I',\n",
       " 'curcumin',\n",
       " 'cyclophosphamide',\n",
       " 'cytarabine hydrochloride',\n",
       " 'cytochalasin B',\n",
       " 'dabrafenib',\n",
       " 'dacarbazine',\n",
       " 'daporinad',\n",
       " 'darinaparsin',\n",
       " 'dasatinib',\n",
       " 'decitabine',\n",
       " 'dexamethasone',\n",
       " 'dinaciclib',\n",
       " 'docetaxel',\n",
       " 'doxorubicin',\n",
       " 'entinostat',\n",
       " 'epigallocatechin-3-monogallate',\n",
       " 'erastin',\n",
       " 'erismodegib',\n",
       " 'etomoxir',\n",
       " 'etoposide',\n",
       " 'fingolimod',\n",
       " 'fluorouracil',\n",
       " 'fluvastatin',\n",
       " 'foretinib',\n",
       " 'fulvestrant',\n",
       " 'fumonisin B1',\n",
       " 'gefitinib',\n",
       " 'gemcitabine',\n",
       " 'gossypol',\n",
       " 'ibrutinib',\n",
       " 'ifosfamide',\n",
       " 'imatinib',\n",
       " 'importazole',\n",
       " 'indisulam',\n",
       " 'isoevodiamine',\n",
       " 'isoliquiritigenin',\n",
       " 'isonicotinohydroxamic acid',\n",
       " 'istradefylline',\n",
       " 'itraconazole',\n",
       " 'lapatinib',\n",
       " 'lenvatinib',\n",
       " 'leptomycin B',\n",
       " 'linifanib',\n",
       " 'linsitinib',\n",
       " 'lomeguatrib',\n",
       " 'lovastatin',\n",
       " 'marinopyrrole A',\n",
       " 'masitinib',\n",
       " 'methotrexate',\n",
       " 'methylstat',\n",
       " 'mitomycin',\n",
       " 'myricetin',\n",
       " 'myriocin',\n",
       " 'nakiterpiosin',\n",
       " 'narciclasine',\n",
       " 'navitoclax',\n",
       " 'necrostatin-1',\n",
       " 'necrostatin-7',\n",
       " 'necrosulfonamide',\n",
       " 'nelarabine',\n",
       " 'neuronal differentiation inducer III',\n",
       " 'niclosamide',\n",
       " 'nilotinib',\n",
       " 'nutlin-3',\n",
       " 'olaparib',\n",
       " 'omacetaxine mepesuccinate',\n",
       " 'ouabain',\n",
       " 'oxaliplatin',\n",
       " 'paclitaxel',\n",
       " 'palmostatin B',\n",
       " 'pandacostat',\n",
       " 'parbendazole',\n",
       " 'parthenolide',\n",
       " 'pazopanib',\n",
       " 'pevonedistat',\n",
       " 'phloretin',\n",
       " 'pifithrin-alpha',\n",
       " 'piperlongumine',\n",
       " 'pluripotin',\n",
       " 'procarbazine',\n",
       " 'prochlorperazine',\n",
       " 'purmorphamine',\n",
       " 'pyrazolanthrone',\n",
       " 'quizartinib',\n",
       " 'regorafenib',\n",
       " 'rigosertib',\n",
       " 'salermide',\n",
       " 'saracatinib',\n",
       " 'selumetinib',\n",
       " 'semagacestat',\n",
       " 'serdemetan',\n",
       " 'sildenafil',\n",
       " 'silmitasertib',\n",
       " 'simvastatin',\n",
       " 'sirolimus',\n",
       " 'sitagliptin',\n",
       " 'skepinone-L',\n",
       " 'sorafenib',\n",
       " 'sotrastaurin',\n",
       " 'spautin-1',\n",
       " 'staurosporine',\n",
       " 'tacedinaline',\n",
       " 'tamatinib',\n",
       " 'tandutinib',\n",
       " 'temozolomide',\n",
       " 'temsirolimus',\n",
       " 'teniposide',\n",
       " 'thalidomide',\n",
       " 'tigecycline',\n",
       " 'tipifarnib-P1',\n",
       " 'tipifarnib-P2',\n",
       " 'tivantinib',\n",
       " 'tivozanib',\n",
       " 'topotecan',\n",
       " 'tosedostat',\n",
       " 'trametinib',\n",
       " 'tretinoin',\n",
       " 'triazolothiadiazine',\n",
       " 'trifluoperazine',\n",
       " 'triptolide',\n",
       " 'tubastatin A',\n",
       " 'valdecoxib',\n",
       " 'vandetanib',\n",
       " 'veliparib',\n",
       " 'vincristine',\n",
       " 'vorapaxar',\n",
       " 'vorinostat',\n",
       " 'zebularine'}"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df['DRUG_NAME']).intersection(set(all_experiment['DRUG_NAME']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>AUC</th>\n",
       "      <th>DATASET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>136918</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>KM12</td>\n",
       "      <td>0.545506</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136918</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>KM12</td>\n",
       "      <td>0.545506</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173561</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>HT55</td>\n",
       "      <td>0.521087</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180192</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>MDST8</td>\n",
       "      <td>0.507119</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34555</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>SW48</td>\n",
       "      <td>0.522487</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139639</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>RKO</td>\n",
       "      <td>0.520711</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139639</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>RKO</td>\n",
       "      <td>0.520711</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320501</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>SW948</td>\n",
       "      <td>0.415284</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320501</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>SW948</td>\n",
       "      <td>0.415284</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335138</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>SW1417</td>\n",
       "      <td>0.561421</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359874</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>T84</td>\n",
       "      <td>0.533860</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118100</th>\n",
       "      <td>veliparib</td>\n",
       "      <td>C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O</td>\n",
       "      <td>SW1463</td>\n",
       "      <td>0.384035</td>\n",
       "      <td>CTRP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        DRUG_NAME                                  SMILES CELL_LINE_NAME  \\\n",
       "136918  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O           KM12   \n",
       "136918  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O           KM12   \n",
       "173561  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O           HT55   \n",
       "180192  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O          MDST8   \n",
       "34555   veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O           SW48   \n",
       "139639  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O            RKO   \n",
       "139639  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O            RKO   \n",
       "320501  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O          SW948   \n",
       "320501  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O          SW948   \n",
       "335138  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O         SW1417   \n",
       "359874  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O            T84   \n",
       "118100  veliparib  C[C@@]1(CCCN1)c1nc2c(cccc2[nH]1)C(N)=O         SW1463   \n",
       "\n",
       "             AUC DATASET  \n",
       "136918  0.545506    CTRP  \n",
       "136918  0.545506    CTRP  \n",
       "173561  0.521087    CTRP  \n",
       "180192  0.507119    CTRP  \n",
       "34555   0.522487    CTRP  \n",
       "139639  0.520711    CTRP  \n",
       "139639  0.520711    CTRP  \n",
       "320501  0.415284    CTRP  \n",
       "320501  0.415284    CTRP  \n",
       "335138  0.561421    CTRP  \n",
       "359874  0.533860    CTRP  \n",
       "118100  0.384035    CTRP  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_experiment[(all_experiment['CELL_LINE_NAME'].isin(colo_ccl)) &\n",
    "                        (all_experiment['DRUG_NAME'] == 'veliparib')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['alpha-lipoic acid',\n",
       " 'DMOG',\n",
       " 'Nelarabine',\n",
       " 'Selisistat',\n",
       " 'WIKI4',\n",
       " 'AZD1208',\n",
       " 'BEN',\n",
       " 'Cisplatin',\n",
       " 'Temozolomide',\n",
       " 'glutathione',\n",
       " 'Alectinib',\n",
       " 'Lenalidomide',\n",
       " 'Veliparib',\n",
       " '5-azacytidine',\n",
       " 'MIM1',\n",
       " 'Staurosporine',\n",
       " 'Cyclophosphamide',\n",
       " 'Y-39983',\n",
       " 'UNC1215',\n",
       " 'Leflunomide',\n",
       " 'BAM7',\n",
       " 'POMHEX',\n",
       " 'Remodelin',\n",
       " 'Tamoxifen',\n",
       " 'Bexarotene',\n",
       " 'Midostaurin',\n",
       " 'Rucaparib',\n",
       " 'Olaparib',\n",
       " 'Sphingosine Kinase 1 Inhibitor II',\n",
       " 'C-75',\n",
       " 'AGI-5198',\n",
       " 'CPI-613',\n",
       " 'Tenovin-6',\n",
       " 'TGX221',\n",
       " 'CCT007093',\n",
       " 'SB52334',\n",
       " 'Quizartinib',\n",
       " 'NSC319726',\n",
       " 'GSK343',\n",
       " 'LY2109761',\n",
       " 'JNK Inhibitor VIII',\n",
       " 'NG-25',\n",
       " 'Nilotinib',\n",
       " 'PFI-1',\n",
       " 'MIRA-1',\n",
       " 'P22077',\n",
       " 'PCI-34051',\n",
       " 'Torin 2',\n",
       " 'torin2',\n",
       " 'Ara-G',\n",
       " 'Imatinib',\n",
       " 'Ruxolitinib',\n",
       " 'A-83-01',\n",
       " 'GNF-2',\n",
       " 'CCT245232',\n",
       " 'CP466722',\n",
       " 'CPI-637',\n",
       " 'N-acetyl cysteine',\n",
       " 'Idelalisib',\n",
       " 'Doramapimod',\n",
       " 'Dacarbazine',\n",
       " 'Tretinoin',\n",
       " 'AZD8186',\n",
       " 'AMG-319',\n",
       " 'PRIMA-1MET',\n",
       " 'AZD5582',\n",
       " 'Omipalisib',\n",
       " 'Gallibiscoquinazole',\n",
       " 'Voxtalisib',\n",
       " 'Niraparib',\n",
       " 'kb NB 142-70',\n",
       " 'Dactolisib',\n",
       " 'I-CBP112',\n",
       " 'T0901317',\n",
       " 'BAY-61-3606',\n",
       " 'Elephantin',\n",
       " 'ETP-45835',\n",
       " 'Fulvestrant',\n",
       " 'Linsitinib',\n",
       " 'Venetoclax',\n",
       " 'XMD8-92',\n",
       " 'RVX-208',\n",
       " 'AZ6102',\n",
       " 'AR-42',\n",
       " 'A-443654',\n",
       " 'Mycophenolic acid',\n",
       " 'QS11',\n",
       " 'SGC-CBP30',\n",
       " 'Cyclopamine',\n",
       " 'JNK-9L',\n",
       " 'AZD2014',\n",
       " 'GSK650394',\n",
       " 'XMD8-85',\n",
       " 'Zoledronate',\n",
       " 'Lestaurtinib',\n",
       " 'Elesclomol',\n",
       " 'Masitinib',\n",
       " 'UNC0379',\n",
       " 'AICA Ribonucleotide',\n",
       " 'CP724714',\n",
       " 'ARRY-520',\n",
       " 'GSK2578215A',\n",
       " 'FTI-277',\n",
       " 'AKT inhibitor VIII',\n",
       " 'Oxaliplatin',\n",
       " 'HG6-64-1',\n",
       " 'ZM447439',\n",
       " 'Pyridostatin',\n",
       " 'NSC-207895',\n",
       " 'Tubastatin A',\n",
       " '5-Fluorouracil',\n",
       " 'SB216763',\n",
       " 'Sinularin',\n",
       " 'LIMK1 inhibitor BMS4',\n",
       " 'CX-5461',\n",
       " 'AZ960',\n",
       " 'A-366',\n",
       " 'IPA-3',\n",
       " 'Dihydrorotenone',\n",
       " 'STF-62247',\n",
       " 'Pazopanib',\n",
       " 'Embelin',\n",
       " 'CI-1040',\n",
       " 'LCL161',\n",
       " 'GNE-317',\n",
       " 'CGP-082996',\n",
       " 'Bicalutamide',\n",
       " 'CZC24832',\n",
       " 'Tivozanib',\n",
       " 'Talazoparib',\n",
       " 'Pilaralisib',\n",
       " 'AZD1332',\n",
       " 'LGK974',\n",
       " 'Acetalax',\n",
       " 'Parthenolide',\n",
       " 'Alpelisib',\n",
       " 'Motesanib',\n",
       " 'Trichostatin A',\n",
       " 'TAK-715',\n",
       " 'Palbociclib',\n",
       " 'NVP-BHG712',\n",
       " 'AZD3514',\n",
       " 'XAV939',\n",
       " 'LJI308',\n",
       " 'S-Trityl-L-cysteine',\n",
       " 'EPZ5676',\n",
       " 'ABT737',\n",
       " 'TWS119',\n",
       " 'Navitoclax',\n",
       " 'Fludarabine',\n",
       " 'SB505124',\n",
       " 'ZG-10',\n",
       " 'GSK2801',\n",
       " 'Cabozantinib',\n",
       " 'Amuvatinib',\n",
       " 'WHI-P97',\n",
       " 'GSK319347A',\n",
       " 'Pyrimethamine',\n",
       " 'Pemetrexed',\n",
       " 'SCH772984',\n",
       " 'WEHI-539',\n",
       " 'GSK-LSD1',\n",
       " 'Ponatinib',\n",
       " 'FR-180204',\n",
       " 'GSK2606414',\n",
       " 'Sunitinib',\n",
       " 'LFM-A13',\n",
       " 'Taselisib',\n",
       " 'UNC0642',\n",
       " 'RO-3306',\n",
       " 'BIX02189',\n",
       " 'AZD6738',\n",
       " 'PF-562271',\n",
       " 'Seliciclib',\n",
       " 'SL0101',\n",
       " 'EHT-1864',\n",
       " 'QL-X-138',\n",
       " 'BMS-509744',\n",
       " 'NU7441',\n",
       " 'Phenformin',\n",
       " 'Wnt-C59',\n",
       " 'GSK591',\n",
       " 'I-BET-762',\n",
       " 'Serdemetan',\n",
       " 'AZ628',\n",
       " 'Tipifarnib',\n",
       " 'Romidepsin',\n",
       " 'EPZ004777',\n",
       " 'SGC0946',\n",
       " 'I-BRD9',\n",
       " '(5Z)-7-Oxozeaenol',\n",
       " 'VX-702',\n",
       " 'BX-912',\n",
       " 'MN-64',\n",
       " 'PF-4708671',\n",
       " 'UMI-77',\n",
       " 'PD0325901',\n",
       " 'KIN001-266',\n",
       " 'Sorafenib',\n",
       " 'CAY10566',\n",
       " 'L-Oxonoreleagnine',\n",
       " 'Ibrutinib',\n",
       " 'Dabrafenib',\n",
       " 'Refametinib',\n",
       " 'AT13148',\n",
       " 'Capivasertib',\n",
       " 'AZD5363',\n",
       " 'JNJ38877605',\n",
       " 'GSK269962A',\n",
       " 'AS601245',\n",
       " 'Osimertinib',\n",
       " 'GSK1070916',\n",
       " 'CD532',\n",
       " 'Lapatinib',\n",
       " 'Podophyllotoxin bromide',\n",
       " 'Axitinib',\n",
       " 'Zibotentan',\n",
       " 'AGK2',\n",
       " 'GW441756',\n",
       " 'Tanespimycin',\n",
       " 'AZD3759',\n",
       " 'Epothilone B',\n",
       " 'WZ4003',\n",
       " 'NSC-87877',\n",
       " 'OF-1',\n",
       " 'Gefitinib',\n",
       " 'Mitomycin-C',\n",
       " 'Bosutinib',\n",
       " 'Entospletinib',\n",
       " 'AZD4877',\n",
       " 'Cytarabine',\n",
       " 'ACY-1215',\n",
       " 'GDC0810',\n",
       " 'Telomerase Inhibitor IX',\n",
       " 'AZD5438',\n",
       " 'OSU-03012',\n",
       " 'Schweinfurthin A',\n",
       " 'Bromosporine',\n",
       " 'Erlotinib',\n",
       " 'VE821',\n",
       " 'Buparlisib',\n",
       " 'MK-8776',\n",
       " 'FMK',\n",
       " 'RU-SKI 43',\n",
       " 'YM201636',\n",
       " 'WH-4-023',\n",
       " 'AT-7519',\n",
       " 'Vinorelbine',\n",
       " 'Z-LLNle-CHO',\n",
       " 'ML323',\n",
       " 'GSK2830371',\n",
       " 'CCT-018159',\n",
       " 'AZ20',\n",
       " 'GSK1904529A',\n",
       " 'SB590885',\n",
       " 'IMD-0354',\n",
       " 'Enzastaurin',\n",
       " 'PHA-665752',\n",
       " 'Sabutoclax',\n",
       " 'Trametinib',\n",
       " 'Salubrinal',\n",
       " 'AST-1306',\n",
       " 'IWP-2',\n",
       " 'A-770041',\n",
       " 'GW843682X',\n",
       " 'Saracatinib',\n",
       " 'Piperlongumine',\n",
       " 'AZD6094',\n",
       " 'Savolitinib',\n",
       " 'Selumetinib',\n",
       " 'Uprosertib',\n",
       " 'Mitoxantrone',\n",
       " 'Thapsigargin',\n",
       " 'Ribociclib',\n",
       " 'YK-4-279',\n",
       " 'Pevonedistat',\n",
       " 'AZD5991',\n",
       " 'CRT0105950',\n",
       " 'PD173074',\n",
       " 'LDN-193189',\n",
       " 'Ipatasertib',\n",
       " 'FTY-720',\n",
       " 'Shikonin',\n",
       " 'AZD8931',\n",
       " 'Sapitinib',\n",
       " 'PFI-3',\n",
       " 'PFI3',\n",
       " 'Bleomycin',\n",
       " 'Afuresertib',\n",
       " 'GSK2110183B',\n",
       " 'Dactinomycin',\n",
       " 'Vorinostat',\n",
       " 'BIBF-1120',\n",
       " 'AGI-6780',\n",
       " 'FH535',\n",
       " 'Apitolisib',\n",
       " 'Alisertib',\n",
       " 'VE-822',\n",
       " 'AZD5153',\n",
       " 'CGP-60474',\n",
       " 'BPTES',\n",
       " 'GW-2580',\n",
       " 'Mirin',\n",
       " 'Linifanib',\n",
       " 'CUDC-101',\n",
       " 'Bryostatin 1',\n",
       " 'Ispinesib Mesylate',\n",
       " 'GSK429286A',\n",
       " 'Ulixertinib',\n",
       " 'CI-1033',\n",
       " 'Avagacestat',\n",
       " 'I-BET-151',\n",
       " 'BX795',\n",
       " 'CAY10603',\n",
       " 'Temsirolimus',\n",
       " 'TW 37',\n",
       " 'Etoposide',\n",
       " 'Kobe2602',\n",
       " 'Rapamycin',\n",
       " 'Daporinad',\n",
       " 'Foretinib',\n",
       " 'Pelitinib',\n",
       " 'PF-00299804',\n",
       " 'Entinostat',\n",
       " 'Pictilisib',\n",
       " 'AS605240',\n",
       " 'Irinotecan',\n",
       " 'Cediranib',\n",
       " 'Bortezomib',\n",
       " 'Tozasertib',\n",
       " 'PRT062607',\n",
       " 'Wee1 Inhibitor',\n",
       " 'WYE-125132',\n",
       " 'JQ1',\n",
       " 'Gemcitabine',\n",
       " 'Docetaxel',\n",
       " 'Dinaciclib',\n",
       " 'CMK',\n",
       " 'VX-11e',\n",
       " 'Camptothecin',\n",
       " 'Crizotinib',\n",
       " 'GSK690693',\n",
       " 'Vismodegib',\n",
       " 'OTX015',\n",
       " 'Belinostat',\n",
       " 'Methotrexate',\n",
       " 'Obatoclax Mesylate',\n",
       " 'Fedratinib',\n",
       " 'Dasatinib',\n",
       " 'Afatinib',\n",
       " 'AZD8835',\n",
       " 'Paclitaxel',\n",
       " 'Sepantronium bromide',\n",
       " 'Vinblastine',\n",
       " 'Flavopiridol',\n",
       " 'Panobinostat',\n",
       " 'LMP744',\n",
       " 'Dacinostat',\n",
       " 'IOX2',\n",
       " 'Topotecan',\n",
       " 'Teniposide',\n",
       " 'Luminespib',\n",
       " 'Doxorubicin',\n",
       " 'Epirubicin']"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_exclued.groupby(['DRUG_NAME']).mean().sort_values(by='AUC_predicted', ascending=True).index)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tf",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "074ed60ff9a97295a85c7cce8cae3388aa83a28770d726c7b567134e875ed814"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
